src/remote_prune

```diff
@@ -0,0 +1,224 @@
+                                "gcp": metaData.get("gcp") or [],
+                            }
+                            currentBuild = Build(
+                                id=build_id,
+                                timestamp=timestamp,
+                                images=images,
+                                arches=arch,
+                            )
+                            match action:
+                                case "cloud-uploads":
+                                    delete_cloud_resource(currentBuild, cloud_config, args.dry_run)
+                                    if not args.dry_run:
+                                        build.setdefault("policy-cleanup", []).append(action)
+                                        buildJsonData["builds"][index] = build
+                                case "build":
+                                    print(f"Will delete everything for {build_id}")
+        # pprint.pprint(buildJsonData['builds'], compact=True)
+        with open(f"builds/builds.json", "w") as json_file:
+            json_file.write(json.dumps(buildJsonData))
+
+def delete_cloud_resource(build, cloud_config, dry_run):
+    print(f"Deleting cloud uploads for {build.id}")
+    errors = []
+    # Unregister AMIs and snapshots
+    for ami in build.images.get("amis", []):
+        region_name = ami.get("name")
+        ami_id = ami.get("hvm")
+        snapshot_id = ami.get("snapshot")
+        if dry_run:
+            print(f"Will delete {ami_id} and {snapshot_id}")
+        else:
+            if ami_id and region_name:
+                try:
+                    deregister_ami(ami_id, region=region_name)
+                except Exception as e:
+                    errors.append(e)
+            if snapshot_id and region_name:
+                try:
+                    delete_snapshot(snapshot_id, region=region_name)
+                except Exception as e:
+                    errors.append(e)
+
+    aliyun = build.images.get("aliyun")
+    if aliyun:
+        region_name = aliyun.get("name")
+        aliyun_id = aliyun.get("hvm")
+        if dry_run:
+            print(f"Will delete {aliyun_id} in {region_name}")
+        else:
+            if region_name and aliyun_id:
+                try:
+                    remove_aliyun_image(aliyun_id, region=region_name)
+                except Exception as e:
+                    errors.append(e)
+
+    gcp = build.images.get("gcp")
+    if gcp:
+        gcp_image = gcp.get("image")
+        json_key = cloud_config.get("gcp", {}).get("json-key")
+        project = cloud_config.get("gcp", {}).get("project")
+        if dry_run:
+            print(f"Will delete {gcp_image}")
+        else:
+            if gcp_image and json_key and project:
+                try:
+                    remove_gcp_image(gcp_image, json_key, project)
+                except Exception as e:
+                    errors.append(e)
```

Comment:
Hmm, one concern here is designing this in a [fail-safe](https://en.wikipedia.org/wiki/Fail-safe) way. So if a new upload type is added and we forget to add support for it in this script, after long enough it could happen that we start deleting those builds and leak those resources. Ideally, this command should fail instead but that would require it to know about clouds it doesn't know about. :)

One straightforward but awkward way to do this is to have an allowlist of all toplevel keys allowed in the `meta.json`. If a toplevel key is in `meta.json` that we don't know about, we fail. Not sure if there's a less awkward way to do this. Ideally, there'd be a way to leverage the schema to help with this.

For example, right now this doesn't handle `azure`, `kubevirt`, `extensions-container`, and `base-oscontainer` (the latter three could have versioned tags that should ideally be pruned). And that'd be OK for really old builds that don't even have those. But without something to cross-check against, there's nothing that stops this code from also deleting builds that do have them.

Reply:
OK, so here's my idea for this: maintain a list of supported (e.g. `amis`) and unsupported keys (e.g. `kubevirt`). The union of these two lists should represent the full set of possible remote resources that a build could have as per the schema.

If a build is up for cloud pruning, we check that none of the unsupported keys are present. If any are, error out.

To ensure that the list remains in sync with the schema, we keep a hash of the schema JSON file as a comment in this file, and add a CI check in this repo that greps for the hash of the schema. This means that any schema change requires changing the comment too. The comment could say something like:

```python
# These lists are up to date as of schema hash
# 4c19aed3b3d84af278780bff63728510bb3e70613e4c4eef8cabd7939eb31bd8. If changing
# this hash, ensure that the list of supported and unsupported artifacts below
# is up to date.
def supported = [...]
def unsupported = [...]
```